nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,372 @@
+#include "tf2_ros/transform_broadcaster.h"
+
+#include "nav2_costmap_2d/costmap_filters/filter_values.hpp"
+#include "nav2_util/occ_grid_values.hpp"
+
+namespace nav2_costmap_2d
+{
+
+SpeedFilter::SpeedFilter()
+: speed_limit_topic_(""), filter_info_sub_(nullptr), mask_sub_(nullptr),
+  speed_limit_pub_(nullptr), filter_mask_(nullptr), mask_frame_(""), global_frame_("")
+{
+}
+
+void SpeedFilter::initializeFilter(
+  const std::string & filter_info_topic)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Declare "speed_limit_topic" parameter specific to SpeedFilter only
+  declareParameter("speed_limit_topic", rclcpp::ParameterValue("speed_limit"));
+  node->get_parameter(name_ + "." + "speed_limit_topic", speed_limit_topic_);
+
+  filter_info_topic_ = filter_info_topic;
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    node->get_logger(),
+    "Subscribing to \"%s\" topic for filter info...",
+    filter_info_topic_.c_str());
+  filter_info_sub_ = node->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    filter_info_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&SpeedFilter::filterInfoCallback, this, std::placeholders::_1));
+
+  // Get global frame required for speed limit publisher
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+
+  // Create new speed limit publisher
+  speed_limit_pub_ = node->create_publisher<nav2_msgs::msg::SpeedLimit>(
+    speed_limit_topic_,
+    rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable());
+  speed_limit_pub_->on_activate();
+
+  // Reset speed conversion states
+  base_ = BASE_DEFAULT;
+  multiplier_ = MULTIPLIER_DEFAULT;
+  percentage_ = true;
+  speed_limit_ = NO_SPEED_LIMIT;
+  speed_limit_prev_ = FAKE_SPEED_LIMIT;  // Fake speed limit to force speed limit topic update
+}
+
+void SpeedFilter::filterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  // Set base_/multiplier_ or use speed limit in % of maximum speed
+  base_ = msg->base;
+  multiplier_ = msg->multiplier;
+  if (msg->type == SPEED_FILTER_ABSOLUTE) {
+    // Using speed limit in m/s
+    percentage_ = false;
```

Comment:
I think the correct thing here is to throw an exception with text that this is not implemented yet.

Reply:
Done in https://github.com/ros-planning/navigation2/pull/2074/commits/400c1200ec5967b34a1e6e9907ae79e1326faaae#diff-d344e376ecbc8e5ecd5045731e13855c6d2a9a955cdaee9b31dddbbee7681466R129